python/ray/experimental/state/state_cli.py

```diff
@@ -37,25 +37,56 @@ class AvailableFormat(Enum):
     TABLE = "table"
 
 
-def _parse_filter(filter: str) -> Tuple[str, AvailablePredicate, SupportedFilterType]:
+def _parse_filter(filter: str) -> Tuple[str, PredicateType, SupportedFilterType]:
     """Parse the filter string to a tuple of key, preciate, and value."""
     # The function assumes there's going to be no key that includes "="" or "!=".
-    if "!=" in filter:
```

Comment:
Therer were actually edge cases from this logic when != and = are included within the same string! (so we should always catch the first = or != predicate from the filter)

Reply:
Oh i see, like this?
```
"key=value!=lol"
```

Just throwing untested ideas here:
If we split by `!=` or `=` with `str.split(<delimiter>, 1)`, and use the delimiter pos that's smaller if present, we should be able to handle such cases?

Reply:
Interesting idea... Didn't know we can choose the max split from `split` API.

```python3
def _parse_filter(filter: str) -> Tuple[str, PredicateType, SupportedFilterType]:
    """Parse the filter string to a tuple of key, preciate, and value."""
    # The function assumes there's going to be no key that includes "="" or "!=".
    # Since key is controlled by us, it should be trivial to keep the invariant.

    predicate = None
    min_index = len(filter)
    predicate_candidates = {
        "=": filter.split("=", 1),
        "!=": filter.split("!=", 1)
    }

    for pred, kv in predicate_candidates.items():
        # If it is splittable by a given filter delimeter.
        if len(kv) == 2:
            k, _ = kv

            # Choose the lowest length key because
            # that's the first found delimeter from the filter.
            # E.g., key=abc!=value -> first found delimeter is `=`
            # which has lower index.
            if len(k) < min_index:
                predicate = pred
                min_index = len(k)

    if not predicate:
        raise ValueError(
            f"The format of a given filter {filter} is invalid: "
            "Cannot find the predicate. "
            "Please provide key=val or key!=val format string."
        )

    k, v = predicate_candidates[predicate]
    if len(k) == 0 or len(v) == 0:
        raise ValueError(
            f"The format of a given filter {filter} is invalid: "
            f"Cannot identify key {k} or value, {v} (it is empty). "
            "Please provide key=val or key!=val format string."
        )

    return (k, predicate, v)
```

The core logic is definitely shorter (which I like). However, I personally felt it needs a bit more comments to explain the logic when I tried. So I will just stick to the original implementation (but feel free to give me simpler code than this...)